[Merged by Bors] - Fix dependency of shadow mapping on the optional `PrepassPlugin`

geieredgar opened the pull request:
# Objective

Unfortunately, there are three issues with my changes introduced by #7784.

1.  The changes left some dead code. This is already taken care of here: #7875.
2. Disabling prepass causes failures because the shadow mapping relies on the `PrepassPlugin` now.
3. Custom materials use the `prepass.wgsl` shader, but this does not always define a fragment entry point.

This PR fixes 2. and 3. and resolves #7879.

## Solution

- Add a regression test with disabled prepass.
- Split `PrepassPlugin` into two plugins:
  - `PrepassPipelinePlugin` contains the part that is required for the shadow mapping to work and is unconditionally added.
  - `PrepassPlugin` now only adds the systems and resources required for the "real" prepasses.
- Add a noop fragment entry point to `prepass.wgsl`, used if `NORMAL_PASS` is not defined.

IceSentry commented on crates/bevy_pbr/src/prepass/mod.rs:
```diff
@@ -104,14 +135,11 @@ where
             .init_resource::<DrawFunctions<Opaque3dPrepass>>()
             .init_resource::<DrawFunctions<AlphaMask3dPrepass>>()
-            .init_resource::<PrepassViewBindGroup>()
             .init_resource::<SpecializedMeshPipelines<PrepassPipeline<M>>>()
```
Isn't this already added by the prepassPipelinePlugin?

geieredgar replied:
Yes, thank you for catching that. I fixed that.

geieredgar commented:
While investigating #7879, I have found a third issue: When using custom materials, the default `prepass.wgsl` shader could be used.  But `prepass.wgsl` only defined an entry point for the fragment shader if `NORMAL_PASS` was defined. I added an entrypoint for the other cases that simply does nothing. This will produce the same shadows like before #7784.  So a custom material should override the `prepass_fragment_shader` method to be able to obtain better shadows like the `StandardMaterial`.

IceSentry commented on crates/bevy_pbr/src/prepass/prepass.wgsl:
```diff
@@ -81,4 +82,10 @@ struct FragmentInput {
+#else // NORMAL_PREPASS
+
+@fragment
+fn fragment(in: VertexOutput) {}
```
I'm not sure I understand why this is necessary. The pipeline sets the fragment state to None if the normal prepass is not enabled.

geieredgar replied:
This was previously the case, yes, although not exactly. Before my changes, the fragment state was set if either the normal prepass was enabled or the material used an alpha mask. I believe, if the `CustomMaterial` in the `shader_prepass` example would have used `AlphaMode::Mask` as `alpha_mode`, it would break even without my changes. 
Now the fragment state will also be set if the alpha mode is `Blend`, `Add` or `Premultiplied`. This triggers the issue with the missing entry point because the `CustomMaterial` uses `AlphaMode::Blend`.

geieredgar replied:
I can confirm my suspicion, going back to a39c22386b0eadd5c39e342a2ce9dd4d920bc492 right before the merge of #7784 and changing the alpha mode to `AlphaMode::Mask(0.5)` causes the example to fail with:
```
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `prepass_pipeline`
    error matching FRAGMENT shader requirements against the pipeline
    location[0] is provided by the previous stage output but is not consumed as input by this stage.

'
```
Interestingly, it doesn't complain about a missing fragment entry point, which is what I expected.

superdump commented on crates/bevy_pbr/src/prepass/mod.rs:
```diff
@@ -59,15 +59,18 @@ pub const PREPASS_BINDINGS_SHADER_HANDLE: HandleUntyped =
-pub struct PrepassPlugin<M: Material>(PhantomData<M>);
+/// Sets up everything required to use the prepass pipeline.
+///
+/// This does not add the actual prepasses, see [`PrepassPipeline`] for that.
```
Is this supposed to refer to PrepassPlugin?
```suggestion
/// This does not add the actual prepasses, see [`PrepassPlugin`] for that.
```

superdump replied:
After this is clarified, LGTM and we can merge.

robtfm reviewed (commented):
This does work, but I don't think adding the dummy fragment shader should be necessary. instead we can just not add the FragmentState in `PrepassPipeline::specialize()` when the `material_fragment_shader` is `None`.

`StandardMaterial` returns a handle, so that will behave correctly. Custom Materials that return None would not add a frag shader, which would be more efficient.

0.10 is looming though, so if there isn't time then this is fine as is

geieredgar commented:
Ok, I adjusted the fragment state condition and its comment, reverted the fragment entry point commit and tested that the `shader_prepass` example still works. I think this PR is ready to be merged.

alice-i-cecile commented:
bors r+
